# Cálculo de raíces cuadradas

Utilizaremos el método de la sucesión convergente para calcular la raíz cuadrada de un número $a\in\mathbb{R}$ positivo:

\begin{gather}
 x_0 \in \mathbb{R}: \; x_0^2\approx a \\
 x_{n+1} = \frac{1}{2}\left( x_n + \frac{a}{x_n} \right)\\
 (\text{Se sabe que}: \lim_{n\to\infty} x_n = \sqrt{a})
\end{gather}

In [3]:
from fractions import Fraction as F

In [4]:
F(2,4)

Fraction(1, 2)

Este módulo es útil para hacer operaciones con fracciones.

In [3]:
x=F(2,4) # 1/2
x

Fraction(1, 2)

Una de sus gracias es que las expresa en forma irreducible.

> **Nota**: Una fracción $\frac{a}{b}$ es irreducible cuando $a$ y $b$ son primos relativos (es decir, el máximo común divisor de $a$ y $b$ es 1).

Además, podemos obtener la representación decimal (_aproximada_) de esas fracciones.

In [4]:
float(x)

0.5

In [5]:
float(F(2,3))

0.6666666666666666

Vamos a definir una función, de forma recursiva, que nos permita obtener el término $x_n$ del listado de números que converge a la raíz de $a$. Para ello, vamos a ejemplificar la definición y uso de las funciones en [`Python 3`](https://www.python.org/).

In [5]:
# Vamos a hacer una función que, dados dos números, calcule su suma
def raul(a,b): #El nombre de la función es arbitrario
    return a+b

raul(10,2.5)

12.5

In [6]:
def memo(papitas,refresco): #También el nombre de los argumentos es arbitrario (aunque se prefiere algo que )
    return papitas*refresco # todavía no haya sido utilizado).

memo(5,2)

10

In [9]:
memo(5,4)

20

Ahora vamos a definir la función que me dé la raíz de un número `a` a partir de una aproximación inicial `x0` hasta un orden `n`:

In [10]:
def raiz(a,x0,n):
    """
    a = número cuya raíz me interesa
    x0 = aproximación inicial de la raíz de a (se recomienda un entero o fracción cuyo cuadrado sea casi a)
    n = número natural para el que calcularemos x1, x2,..., xn
    """
    y=x0
    for i in range(n):
        print(f'X({i})={y}\n     ={float(y)}\n')
        y=F(1,2)*(y+F(a,y)) # Esto codifica la fórmula recursiva definida al inicio
    print(f'X({n})={y}\n     ={float(y)}')
    return y

In [11]:
y=raiz(10,3,5)

X(0)=3
     =3.0

X(1)=19/6
     =3.1666666666666665

X(2)=721/228
     =3.162280701754386

X(3)=1039681/328776
     =3.162277660169842

X(4)=2161873163521/683644320912
     =3.1622776601683795

X(5)=9347391150304592810234881/2955904621546382351702304
     =3.1622776601683795


La aproximación obtenida para la raíz de 10 es tal, que el cuadrado _se pasa de 10_ en:

In [14]:
e=y**2-10
print(e)

1/8737372131679261877757908894145308222766638908416


In [16]:
float(e)

1.1445088808502047e-49

Este número está expresado en notación científica (computacionalmente):

$$ \text{1.1445088808502047e-49} = 1.1445088808502047 \times 10^{-49} = \\
0.00000000000000000000000000000000000000000000000011445088808502047
$$

> **Nota**: Lo más chiquito que una computadora puede procesar se llama [_épsilon de máquina_](https://en.wikipedia.org/wiki/Machine_epsilon).

# Cuestionario

En equipos, responder (utilizando los cuadernillos de `Jupyter`):

1. ¿Cuál es la raíz de 9 a orden 3, 4 y 5 para la aproximación inicial $x_0=4$?
2. ¿Qué pasa con la raíz de 25 a órdenes 2, 3 y 4 cuando $x_0=5$?
3. ¿Cuál es el error del cuadrado de $x_3$ y $x_5$ para la raíz de 11 cuando $x_0=5$?

1.

In [14]:
raiz(9,4,5)
#9 es el número cuya raíz nos interesa. 4 es la aproximación inicial. 5 es el elemento de la sucesión que nos interesa.

X(0)=4
     =4.0

X(1)=25/8
     =3.125

X(2)=1201/400
     =3.0025

X(3)=2882401/960800
     =3.000001040799334

X(4)=16616465284801/5538821761600
     =3.0000000000001807

X(5)=552213837121960323152649601/184071279040653441050883200
     =3.0


Fraction(552213837121960323152649601, 184071279040653441050883200)

In [16]:
_-3

Fraction(1, 184071279040653441050883200)

2.

In [17]:
raiz(25,5,4)
#9 es el número cuya raíz nos interesa. 5 es la aproximación inicial. 4 es el elemento de la sucesión que nos interesa.

X(0)=5
     =5.0

X(1)=5
     =5.0

X(2)=5
     =5.0

X(3)=5
     =5.0

X(4)=5
     =5.0


Fraction(5, 1)

3. ¿Cuál es el error del cuadrado de $x_3$ y $x_5$ para la raíz de 11 cuando $x_0=5$?

In [19]:
x3=raiz(11,5,3)

X(0)=5
     =5.0

X(1)=18/5
     =3.6

X(2)=599/180
     =3.327777777777778

X(3)=715201/215640
     =3.3166434798738638


In [23]:
#El error del cuadrado de x3 para la raíz de 11 cuando x0=5
float(x3**2)-11

0.00012397258981344805

In [20]:
x5=raiz(11,5,5)

X(0)=5
     =5.0

X(1)=18/5
     =3.6

X(2)=599/180
     =3.327777777777778

X(3)=715201/215640
     =3.3166434798738638

X(4)=1023019176001/308451887280
     =3.316624790408058

X(5)=2093136468898297098134401/631104391122277866334560
     =3.3166247903554


In [24]:
#El error del cuadrado de x5 para la raíz de 11 cuando x0=5
float(x5**2)-11

0.0

## <font color='crimson'> Participantes </font>

* Angejolie González
* José Luis Carrizosa
* Omar Santiago Martínez González
* Aaron Fernando Del Valle Reyes
* González Santana Abraham Alejandro (+1)